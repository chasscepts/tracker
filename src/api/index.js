import axios from 'axios';

const normalizeError = (err) => {
  if (!err) {
    return { message: 'An unknown error encountered. Please try again.' };
  }
  if (err.response) {
    return { message: err.response.data.message || JSON.stringify(err.response.data) };
  }
  if (err.request) {
    return { message: 'Server is not responding.' };
  }
  if (err.message) {
    return { message: err.message };
  }
  if (typeof err === 'string') {
    return { message: err };
  }
  return { message: 'An unknown error encountered. Please try again.' };
};

const instantiate = (token) => {
  const instance = axios.create({
    baseURL: 'http://localhost:3000/',
    responseType: 'json',
  });
  instance.defaults.headers.common.Authorization = `Bearer ${token}`;

  return instance;
};

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const get = (instance, path) => new Promise((resolve, reject) => {
  instance.get(path)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

/**
 *
 * @param {axios} instance An instance of axios to use for this request
 * @param {string} path relative url
 * @param {{ string: any }} data sent in body of this post
 * @returns Promise that resolves to fetched data when request is successful
 * and rejects with error when request fails
 */
const post = (instance, path, data) => new Promise((resolve, reject) => {
  instance.post(path, data)
    .then((res) => resolve(res.data))
    .catch((err) => reject(normalizeError(err)));
});

const api = {
  getGroups: (token) => get(instantiate(token), '/groups'),
  getTasks: (token, start, end) => {
    let path = '/tasks';
    if (start) {
      path = `${path}?start=${start}`;
      if (end) {
        path = `${path}&end=${end}`;
      }
    }

    return get(instantiate(token), path);
  },
  updateTask: (token, title) => post(instantiate(token), '/tasks', { title }),
};

export default api;
