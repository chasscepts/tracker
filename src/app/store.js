import { configureStore } from '@reduxjs/toolkit';
import authReducer from '../reducers/authSlice';
import tasksReducer from '../reducers/tasksSlice';
import timerReducer from '../reducers/timerSlice';

export default configureStore({
  reducer: {
    auth: authReducer,
    tasks: tasksReducer,
    timer: timerReducer,
  },
});
