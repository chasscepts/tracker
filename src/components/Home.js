import DatePanel from './DatePanel';
import GroupPanel from '../containers/GroupPanel';
import TaskPanel from '../containers/TaskPanel';

export default function Home() {
  return (
    <>
      <DatePanel />
      <GroupPanel />
      <TaskPanel />
    </>
  );
}
