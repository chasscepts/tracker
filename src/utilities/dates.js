const doubleDigits = (num) => {
  if (num < 10) return `0${num}`;
  return num;
};

/**
 * @param {Date} date  Date object
 * @returns string representation of date that conforms to API format
 */
const formatDate = (date) => `${date.getFullYear()}-${doubleDigits(date.getMonth() + 1)}-${doubleDigits(date.getDate())}`;

/**
 * @returns string resentation of today that conforms to API format
 */
export const today = () => formatDate(new Date());

export const format = formatDate;

export const timeStringClock = (time) => {
  let rem = time;
  const hrs = Math.floor(rem / 3600);
  rem -= 3600 * hrs;
  const mins = Math.floor(rem / 60);
  const secs = Math.floor(rem - (60 * mins));
  return `${doubleDigits(hrs)} : ${doubleDigits(mins)} : ${doubleDigits(secs)}`;
};

export const timeString = (time) => {
  let rem = time;
  const hrs = Math.floor(rem / 3600);
  rem -= 3600 * hrs;
  const mins = Math.floor(rem / 60);
  const secs = Math.floor(rem - (60 * mins));
  const ch = hrs === 1 ? 'hr' : 'hrs';
  const cm = mins === 1 ? 'min' : 'mins';
  const cs = secs === 1 ? 'sec' : 'secs';
  return `${hrs}${ch} ${mins}${cm} ${secs}${cs}`;
};

export const toDate = (str) => {
  const parts = str.split('-');
  return new Date(parts[0], parts[1] - 1, parts[2]);
};

/**
* Gets the date of {@num} number of days ago
* @param {*} num number of days in the past
* @returns string resentation of day that conforms to API format
*/
export const daysAgo = (num) => {
  const date = new Date();
  return formatDate(new Date(date.getFullYear(), date.getMonth(), date.getDate() - num));
};

export const formattedDate = () => {
  const today = new Date();
  return today.toLocaleDateString(
    'en-US',
    { year: 'numeric', month: 'long', day: 'numeric' },
  );
};
