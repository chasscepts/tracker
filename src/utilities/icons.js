import study from '../assets/images/school.png';
import office from '../assets/images/office.png';
import meals from '../assets/images/coffee.png';
import sleep from '../assets/images/night.png';
import exercise from '../assets/images/sneaker.png';
import game from '../assets/images/game.png';
import task from '../assets/images/task.png';

const icons = (title) => {
  switch (title) {
    case 'Study':
      return study;
    case 'Office':
      return office;
    case 'Meals':
      return meals;
    case 'Sleep':
      return sleep;
    case 'Exercise':
      return exercise;
    case 'Games':
      return game;
    default:
      return task;
  }
};

export default icons;
