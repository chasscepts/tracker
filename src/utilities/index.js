const doubleDigits = (num) => {
  if (num < 10) return `0${num}`;
  return num;
};

/**
 * @param {Date} date  Date object
 * @returns string representation of date that conforms to API format
 */
const formatDate = (date) => `${date.getFullYear()}-${doubleDigits(date.getMonth() + 1)}-${doubleDigits(date.getDate())}`;

export const COLORS = {
  primary: '#62b5e5',
  gray: '#67899c',
  red: '#9e0606',
};

export const dates = {
  /**
   * @returns string resentation of today that conforms to API format
   */
  today: () => formatDate(new Date()),
  format: formatDate,
  timeStringClock: (time) => {
    let rem = time;
    const hrs = Math.floor(rem / 3600);
    rem -= 3600 * hrs;
    const mins = Math.floor(rem / 60);
    const secs = Math.floor(rem - (60 * mins));
    return `${doubleDigits(hrs)} : ${doubleDigits(mins)} : ${doubleDigits(secs)}`;
  },
  timeString: (time) => {
    let rem = time;
    const hrs = Math.floor(rem / 3600);
    rem -= 3600 * hrs;
    const mins = Math.floor(rem / 60);
    const secs = Math.floor(rem - (60 * mins));
    const ch = hrs === 1 ? 'hr' : 'hrs';
    const cm = mins === 1 ? 'min' : 'mins';
    const cs = secs === 1 ? 'sec' : 'secs';
    return `${hrs}${ch} ${mins}${cm} ${secs}${cs}`;
  },
  toDate: (str) => {
    const parts = str.split('-');
    return new Date(parts[0], parts[1] - 1, parts[2]);
  },
};

export const taskDuration = (task, date) => {
  const entry = task.entries.find((entry) => entry.entry_date === date);
  return (entry && entry.duration) || 0;
};
