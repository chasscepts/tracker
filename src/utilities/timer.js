let elapsed = 0;
let seconds = 0;
let stopped = true;
let prevStamp;
let client;

let id = 0;

let listeners = [];

const subscribe = (callback) => {
  const temp = id;
  id += 1;
  listeners.push({ id, callback });
  return temp;
};

const unsubscribe = (id) => {
  listeners = listeners.filter((l) => l.id !== id);
};

const raiseSecondElapsed = (elapsed) => {
  listeners.forEach((l) => l.callback(elapsed));
};

const loop = (timestamp) => {
  if (stopped) return;
  if (prevStamp === undefined) {
    prevStamp = timestamp;
  } else {
    elapsed += timestamp - prevStamp;
    prevStamp = timestamp;
    const temp = Math.floor(elapsed / 1000);
    if (temp !== seconds) {
      seconds = temp;
      raiseSecondElapsed(temp);
    }
  }
  requestAnimationFrame(loop);
};

const stop = () => {
  stopped = true;
  prevStamp = undefined;
  client = null;
};

const start = (params) => {
  client = params;
  elapsed = 0;
  seconds = 0;
  stopped = false;
  requestAnimationFrame(loop);
};

const resume = () => {
  stopped = false;
  requestAnimationFrame(loop);
};

export default {
  start,
  stop,
  resume,
  subscribe,
  unsubscribe,
  running: () => !stopped,
  elapsedTime: () => seconds,
  getClient: () => client,
};
