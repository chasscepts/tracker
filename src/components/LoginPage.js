import { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Redirect, useLocation } from 'react-router';
import { Link } from 'react-router-dom';
import mb from '../assets/css/MobileInput.module.css';
import {
  selectLoginError, login, selectUser, clearLoginError,
} from '../reducers/authSlice';
import { COLORS } from '../utilities';
import LoadingButton from './LoadingButton';

const styles = {
  container: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    padding: '30px',
  },
  innerWrap: {
    position: 'relative',
    padding: '15px',
    border: '1px solid #ddd',
    borderRadius: '8px',
  },
  h2: {
    color: COLORS.primary,
  },
  controls: {
    marginTop: '30px',
    display: 'flex',
    justifyContent: 'flex-end',
  },
  link: {
    color: '#9e0606',
    textDecoration: 'underline',
  },
  error: {
    fontSize: '0.8rem',
    fontWeight: 'bold',
    color: 'red',
    textAlign: 'center',
    padding: '15px 0',
  },
};

export default function LoginPage() {
  const [email, setEmail] = useState('');
  const [password, setPassord] = useState('');
  const [localError, setLocalError] = useState(null);
  const [busy, setBusy] = useState(false);
  const error = useSelector(selectLoginError);
  const user = useSelector(selectUser);
  const state = useLocation();
  const dispatch = useDispatch();

  if (user) {
    return <Redirect to={state?.from || '/'} />;
  }
  if (error) {
    if (busy) {
      setBusy(false);
    }
    if (error !== localError) {
      setLocalError(error);
    }
    setTimeout(() => dispatch(clearLoginError()), 0);
  }

  const handleTextChange = (evt) => {
    const { name, value } = evt.target;
    if (name === 'email') {
      setEmail(value);
    } else if (name === 'password') {
      setPassord(value);
    }
  };

  const handleSubmit = () => {
    if (email && password) {
      setBusy(true);
      dispatch(login(email, password));
    }
  };

  return (
    <div style={styles.container}>
      <div style={styles.innerWrap}>
        <h2 style={styles.h2}>Sign In</h2>
        {localError && <div style={styles.error}>{localError}</div>}
        <input className={mb.text} type="text" name="email" value={email} placeholder="Enter Email" onChange={handleTextChange} />
        <input className={mb.text} type="password" name="password" value={password} placeholder="Enter Password" onChange={handleTextChange} />
        <LoadingButton label="Log In" onClick={handleSubmit} loading={busy} />
        <div style={styles.controls}>
          <Link to="/register" style={styles.link}>Register</Link>
        </div>
      </div>
    </div>
  );
}
