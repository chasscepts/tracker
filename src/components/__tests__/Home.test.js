import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Home from '../Home';
import { fetchGroups, fetchTasks, fetchUser } from '../../api';
import { formattedDate } from '../../utilities/dates';
import store from '../../app/store';
import LoginPage from '../LoginPage';

jest.mock('../../api');
jest.mock('../../utilities/dates');

const groups = [{ id: 1, title: 'Group 1' }];

const tasks = [
  {
    id: 1,
    title: 'Task 1',
    group_id: 1,
    user_id: 1,
    entries: [
      { id: 1, task_id: 1, entry_date: '2021-09-21', duration: 0 },
    ],
  },
];

function Wrapper() {
  return (
    <Provider store={store}>
      <MemoryRouter>
        <Home />
      </MemoryRouter>
    </Provider>
  );
}

const login = async () => {
  if (store.getState().auth.user) return Promise.resolve();

  fetchUser.mockResolvedValue({ user: { email: 'test@example.com', token: '111111111' } });
  render(
    <Provider store={store}>
      <MemoryRouter>
        <LoginPage />
      </MemoryRouter>
    </Provider>
  );

  userEvent.type(screen.getByPlaceholderText('Enter Email'), 'test@example.com');
  userEvent.type(screen.getByPlaceholderText('Enter Password'), 'password');
  userEvent.click(screen.getByText('Log In'));

  await Promise.resolve();

  //  await waitFor(() => expect(fetchUser).toHaveBeenCalledTimes(1));
}

describe('Home', () => {

  it('matches snapshot', () => {
    formattedDate.mockReturnValue('September 21, 2021');
    const tree = renderer
      .create(<Wrapper />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays the date', () => {
    formattedDate.mockReturnValue('September 21, 2021');
    render(<Wrapper />);
    expect(screen.getByText('September 21, 2021')).toBeInTheDocument();
  });

  describe('api calls', () => {
    beforeEach(async () => {
    });

    it('displays the groups links', async () => {
      fetchGroups.mockResolvedValue(groups);
      fetchTasks.mockResolvedValue(tasks);
      await login();
      render(<Wrapper />);
      await Promise.resolve();
      groups.forEach((group) => expect(screen.getByText(group.title)).toBeInTheDocument());
    });
  
    it('displays the tasks links', async () => {
      fetchGroups.mockResolvedValue(groups);
      fetchTasks.mockResolvedValue(tasks);
      await login();
      render(<Wrapper />);
      await Promise.resolve();
      tasks.forEach((task) => expect(screen.getByText(task.title)).toBeInTheDocument());
    });
  });
});
