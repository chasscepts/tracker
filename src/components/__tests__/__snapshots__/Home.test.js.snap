// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home matches snapshot 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "height": "100%",
      "overflow": "hidden",
    }
  }
>
  <div
    style={
      Object {
        "padding": "15px",
        "textAlign": "center",
        "width": "100%",
      }
    }
  >
    September 21, 2021
  </div>
  <div
    style={
      Object {
        "alignItems": "center",
        "display": "flex",
        "flexDirection": "column",
        "justifyContent": "center",
        "minHeight": "180px",
        "width": "100%",
      }
    }
  >
    <div
      className="ldsRing"
      style={
        Object {
          "height": "40px",
          "width": "40px",
        }
      }
    >
      <div
        style={
          Object {
            "borderColor": "#62b5e5 transparent transparent transparent",
            "borderWidth": "4px",
            "height": "40px",
            "width": "40px",
          }
        }
      />
      <div
        style={
          Object {
            "borderColor": "#62b5e5 transparent transparent transparent",
            "borderWidth": "4px",
            "height": "40px",
            "width": "40px",
          }
        }
      />
      <div
        style={
          Object {
            "borderColor": "#62b5e5 transparent transparent transparent",
            "borderWidth": "4px",
            "height": "40px",
            "width": "40px",
          }
        }
      />
      <div
        style={
          Object {
            "borderColor": "#62b5e5 transparent transparent transparent",
            "borderWidth": "4px",
            "height": "40px",
            "width": "40px",
          }
        }
      />
    </div>
    <div
      style={
        Object {
          "marginTop": "30px",
          "textAlign": "center",
        }
      }
    >
      Loading groups ...
    </div>
  </div>
  <div
    style={
      Object {
        "flex": 1,
        "overflow": "auto",
      }
    }
  >
    <div
      className="container"
    />
  </div>
</div>
`;
